control-plane/agents/src/bin/core/controller/reconciler/volume/garbage_collector.rs

```diff
@@ -238,6 +248,105 @@ async fn disown_unused_replicas(
     GarbageCollector::squash_results(results)
 }
 
+/// After republishing if the nexus was not shutdown gracefully, we should disown its local
+/// child as that is non reservable.
+#[tracing::instrument(level = "debug", skip(context, volume), fields(volume.uuid = %volume.uuid(), request.reconcile = true))]
+async fn disown_non_reservable_replicas(
+    volume: &mut OperationGuardArc<VolumeSpec>,
+    context: &PollContext,
+) -> PollResult {
+    let target = match context.specs().get_volume_target_nexus(volume.as_ref()) {
+        Some(target) => target.operation_guard()?,
+        None => {
+            // if the volume is not published, then leave the replicas around as they might
+            // still reappear as online by the time we publish
+            return PollResult::Ok(PollerState::Busy);
+        }
+    };
+    if !target.as_ref().status().created()
+        || target.as_ref().is_shutdown()
+        || target.as_ref().dirty()
+    {
+        // don't attempt to disown the replicas if the nexus that should own them is not stable
+        return PollResult::Ok(PollerState::Busy);
+    }
+
+    let volume_state = context.registry().get_volume_state(volume.uuid()).await?;
+    if matches!(
+        volume_state.status,
+        VolumeStatus::Faulted | VolumeStatus::Unknown
+    ) {
+        // don't attempt to disown the replicas if the volume state is faulted or unknown
+        return PollResult::Ok(PollerState::Busy);
+    }
+
+    let mut nexus_info = None;
+    let mut results = vec![];
+
+    // Fetch all nexuses that are not properly shutdown
+    let shutdown_pending_nexuses: Vec<NexusSpec> = context
+        .registry()
+        .specs()
+        .get_volume_failed_shutdown_nexuses(volume.uuid())
+        .await
+        .iter()
+        .map(|nexus| nexus.lock().clone())
+        .collect();
+
+    // Remove the local child from the shutdown pending nexuses as they are
+    // non reservable.
+    for nexus in shutdown_pending_nexuses {
+        let children = nexus.children;
+        for child in children {
+            if child.uri().is_local() {
+                if let Some(uuid) = child.uri().uuid_str() {
+                    if let Ok(replica_id) = ReplicaId::try_from(uuid) {
```

Comment:
I think it's better here to use `as_replica` from the child itself, and then check if the uri if it's local because that contains the replicaId already so we don't need conversions.

Reply:
Yeah right. Thanks